Send NDP NA message upon assigning egress IP

wenqiq opened the pull request:
Send NDP NA message upon assigning egress IP

Implement Neighbor Discovery Protocol(NDP) neighbor advertisement for Egress IPv6 support.
Once an IPv6 IP address has been assigned to Node, an unsolicited Neighbor Advertisement ICMPv6
multicast packet will be sent, announcing the IP to all IPv6 nodes as per RFC4861.

Signed-off-by: Wenqi Qiu <wenqiq@vmware.com>

For #2128

tnqn commented on pkg/agent/util/ndp/ndp.go:
```diff
@@ -0,0 +1,299 @@
+	HopLimit = 255
+)
+
+type INDPAdvertiser interface {
```
Not sure if it's overkill to have an interface, a few structs, a function to do this. For IPv4, it's just a function. Does the struct provide some cache or save some repeated work when sending multiple NDP packets? If not, could we start from a function?

wenqiq replied:
ipv6.PacketConn can be reused when sending multiple NDP packets.

tnqn reviewed (commented):
Thanks @wenqiq. Want to check with you if it can be simplified.

wenqiq commented on pkg/agent/util/ndp/ndp.go:
```diff
@@ -0,0 +1,299 @@
+	return nil
+}
+
+func GratuitousNDPOverIface(srcIP net.IP, iface *net.Interface) error {
```
The function sames like Gratuitous ARP in ipv4,  maybe more proper to call NDPNeighborAdvertiseOverIface.

wenqiq commented on go.mod:
```diff
@@ -3,6 +3,7 @@ module antrea.io/antrea
 go 1.15
 
 require (
+	bou.ke/monkey v1.0.2
```
Got it, Thanks for reminding

wenqiq commented:
@tnqn @xliuxu I updated the patch and used a function to implement NDP neighbor advertisement. PTAL. Thanks.

tnqn commented on pkg/agent/controller/egress/ipassigner/ip_assigner_linux.go:
```diff
@@ -26,10 +25,10 @@ import (
 	"antrea.io/antrea/pkg/agent/util"
 	"antrea.io/antrea/pkg/agent/util/arping"
+	"antrea.io/antrea/pkg/agent/util/ndp"
+	ip2 "antrea.io/antrea/pkg/util/ip"
```
ip2 is not a good alias. But I think you don't really need to import the package. See comment below.

wenqiq replied:
done.
Thanks for reviewing. Some comments have addressed. Still in progress, not completely processed.

tnqn commented on pkg/agent/controller/egress/ipassigner/ip_assigner_linux.go:
```diff
@@ -117,7 +128,7 @@ func (a *ipAssigner) AssignIP(ip string) error {
 		}
 
-		addr := netlink.Addr{IPNet: &net.IPNet{IP: parsedIP, Mask: net.CIDRMask(32, 32)}}
+		addr := netlink.Addr{IPNet: &net.IPNet{IP: parsedIP, Mask: net.CIDRMask(maskLen, maskLen)}}
```
You can use `netlink.NewIPNet(parsedIP)`

wenqiq replied:
Got it, done

tnqn commented on pkg/agent/util/ndp/doc.go:
```diff
@@ -0,0 +1,16 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package ndp contains functions to send NDP advertisement on Linux
```
```suggestion
// Package ndp contains functions to send NDP advertisement on Linux.
```

wenqiq replied:
done

tnqn commented on pkg/agent/util/ndp/ndp_linux.go:
```diff
@@ -0,0 +1,161 @@
+
+const (
+	// Option Length, 8-bit unsigned integer. The length of the option (including the type and length fields) in units of 8 octets.
+	// The value 0 is invalid. Nodes MUST silently discard an ND packet that contains an option with length zero.
```
```suggestion
	// The value 0 is invalid. Nodes MUST silently discard a ND packet that contains an option with length zero.
```

wenqiq replied:
done

tnqn commented on pkg/agent/util/ndp/ndp_linux.go:
```diff
@@ -0,0 +1,161 @@
+	return nil
+}
+
+// NeighborAdvertisement sends an NDP Neighbor Advertisement over interface 'iface' from 'srcIP'.
```
```suggestion
// NeighborAdvertisement sends a NDP Neighbor Advertisement over interface 'iface' from 'srcIP'.
```

wenqiq replied:
done

tnqn commented on pkg/agent/util/ndp/ndp_linux.go:
```diff
@@ -0,0 +1,161 @@
+		return fmt.Errorf("interface address error: %v", err)
+	}
+
+	ipAddr := &net.IPAddr{}
```
```suggestion
	var ipAddr *net.IPAddr
```
Since you create a new instance in the loop.

wenqiq replied:
done

tnqn commented on pkg/agent/util/ndp/ndp_linux.go:
```diff
@@ -0,0 +1,161 @@
+		}
+	}
+
+	ic, err := icmp.ListenPacket("ip6:ipv6-icmp", ipAddr.String())
```
why it needs to listen incoming icmp packets? I thought it just sends an advertisement?

wenqiq replied:
done, use syscall.Socket to send a NA message
`syscall.Socket(syscall.AF_INET6, syscall.SOCK_RAW, syscall.IPPROTO_ICMPV6)`

tnqn commented on pkg/agent/util/ndp/ndp_linux.go:
```diff
@@ -0,0 +1,161 @@
+	optionsBytes, err := marshall()
+	if err != nil {
+		return nil, err
+	}
```
It seems the `marshall` function is pointless, which adds 6 extra lines of code but neither improve the readability much (the whole function is about creating a message, why the options bytes are extracted specially with a generic name "marshall") nor be shared.

wenqiq replied:
done

tnqn commented on pkg/agent/util/ndp/ndp_linux.go:
```diff
@@ -0,0 +1,161 @@
+	}
+
+	cm := &ipv6.ControlMessage{
+		HopLimit: hopLimit,
```
why the hopLimit is set multiple times? I assume only one takes effect.

wenqiq replied:
done

tnqn commented on pkg/agent/util/ndp/ndp_linux.go:
```diff
@@ -0,0 +1,161 @@
+
+	cm := &ipv6.ControlMessage{
+		HopLimit: hopLimit,
+		Src:      ipAddr.IP,
```
If we don't set src, will the OS pick one from the interface?

wenqiq replied:
updated, PTAL.

wenqiq commented:
Do you have any comments or considerations about this PR? @tnqn

tnqn commented on pkg/agent/controller/egress/ipassigner/ip_assigner_linux.go:
```diff
@@ -129,14 +127,16 @@ func (a *ipAssigner) AssignIP(ip string) error {
 		klog.V(2).InfoS("Sent gratuitous ARP", "ip", parsedIP)
-	} else {
-		klog.ErrorS(ipv6NotSupportErr, "Failed to send Advertisement", "ip", parsedIP)
+	} else if addr.IP.To16() != nil {
```
as we have verified the IP, maybe just use "else" to avoid a repeated conversion.

wenqiq replied:
Done

> I assume you have tested this PR in a IPv6 cluster and it worked? Have you checked whether the generated NDP packet can fresh an external Node's ip neighbor cache?

I have tested and I will put some test data here soon.

wenqiq replied:
The external Node can receive the NDP NA or GratuitousARP packets, but IP neighbor cache didn't fresh.

tnqn replied:
IIRC, receiving GARP will only flush the stale entry and won't add a new entry, is this your observation? or the stale entry was not deleted? You mean both NDP NA and GARP didn't work?

wenqiq replied:
I think both NDP NA and GARP can work, sent message packets successfully. The integration test seems back to normal. @tnqn 

In both IPv4 and IPv6 cluster it will add a new entry finally: 
```
ip neigh show
2620:124:6020:1006:250:56ff:fea8:aaaa dev ens192  FAILED
```
ipv4
```
ip neigh show
10.176.26.123 dev ens192 lladdr 00:50:56:a7:36:c1 STALE
```

'10.176.26.123' is an IPv4 address assigned to an Egress Node.

"2620:124:6020:1006:250:56ff:fea8:aaaa" is an IPv6 address assigned to an Egress Node.

In IPv6 cluster, 
k get eg
```
NAME              EGRESSIP                                AGE   NODE
egress-web-ipv6   2620:124:6020:1006:250:56ff:fea8:aaaa   12m   wenqiq01-0
```

The IP has been assigned to Node 'wenqiq01-0'.

We can seen other Node received neighbor advertisement messages:

tcpdump -i ens192  icmp6
```
06:51:25.012447 IP6 fe80::250:56ff:fea7:36c1 > ip6-allnodes: ICMP6, neighbor advertisement, tgt is 2620:124:6020:1006:250:56ff:fea8:aaaa, length 32
06:51:25.013613 IP6 fe80::250:56ff:fea7:36c1 > ip6-allnodes: ICMP6, neighbor advertisement, tgt is 2620:124:6020:1006:250:56ff:fea8:aaaa, length 32
06:51:25.031403 IP6 fe80::250:56ff:fea7:36c1 > ip6-allnodes: ICMP6, neighbor advertisement, tgt is 2620:124:6020:1006:250:56ff:fea8:aaaa, length 32
```

ndp -i ens192 -a global listen
```
ndp listen> neighbor advertisement from fe80::250:56ff:fea7:36c1:
  - router:         false
  - solicited:      false
  - override:       true
  - target address: 2620:124:6020:1006:250:56ff:fea8:aaaa
  - options:
    - target link-layer address: 00:50:56:a7:36:c1
```

Ping from Pod(which has been selected by Egress podSelector) to external Node: 

tcpdump -i ens192  icmp6
```
06:53:05.157995 IP6 2620:124:6020:1006:250:56ff:fea8:aaaa > wenqiq01-2: ICMP6, echo request, seq 3, length 64
06:53:06.171124 IP6 wenqiq01-2 > ff02::1:ffa8:aaaa: ICMP6, neighbor solicitation, who has 2620:124:6020:1006:250:56ff:fea8:aaaa, length 32
06:53:06.181933 IP6 2620:124:6020:1006:250:56ff:fea8:aaaa > wenqiq01-2: ICMP6, echo request, seq 4, length 64
06:53:07.206164 IP6 2620:124:6020:1006:250:56ff:fea8:aaaa > wenqiq01-2: ICMP6, echo request, seq 5, length 64
06:53:07.206251 IP6 wenqiq01-2 > ff02::1:ffa8:aaaa: ICMP6, neighbor solicitation, who has 2620:124:6020:1006:250:56ff:fea8:aaaa, length 32
06:53:08.219153 IP6 wenqiq01-2 > ff02::1:ffa8:aaaa: ICMP6, neighbor solicitation, who has 2620:124:6020:1006:250:56ff:fea8:aaaa, length 32
```

tnqn commented on pkg/agent/util/ndp/ndp_linux_test.go:
```diff
@@ -0,0 +1,65 @@
+	}
+}
+
+func mustIPv6(s string) (net.IP, error) {
```
Maybe put it to pkg/util/ip/ip.go like https://github.com/antrea-io/antrea/blob/fe4457f8bf7082bb927927b279ec17d0423b3766/pkg/util/ip/ip.go#L174-L181 
It can be used in other cases too. I think you don't need to check the result since it's only used when we know the IP must be valid.

wenqiq replied:
Done

tnqn commented on test/integration/agent/ip_assigner_linux_test.go:
```diff
@@ -41,11 +43,17 @@ func TestIPAssigner(t *testing.T) {
+		if err != nil {
+			t.Logf("List ip addr error: %v", err)
+		}
+		assert.NoError(t, errAssign, fmt.Sprintf("Failed to assign a valid IP, ip addrs: %s", string(out)))
```
Why the extra step? Even if it's needed, it uses out when there is an error, then should it only get the ouput when it errors?

wenqiq replied:
Add `IP addr` cmd output to print some test log when testing failed and make debugging easier.

tnqn reviewed (commented):
I assume you have tested this PR in a IPv6 cluster and it worked? Have you checked whether the generated NDP packet can fresh an external Node's ip neighbor cache?

tnqn reviewed (dismissed):
LGTM

tnqn commented:
integration test failed:
```
failed to send neighbor advertisement: new NDP Neighbor Advertisement Message error: hardwareAddr length error: 
```

tnqn commented:
It seems that the code for IPv4 is also wrong as `HardwareAddr` is always empty if the dummy device is newly created.

wenqiq commented:
> It seems that the code for IPv4 is also wrong as `HardwareAddr` is always empty if the dummy device is newly created.

In the test case we use local IP "127.0.0.1" to get IPNetDevice and the interface doesn't have a HardwareAddr which cause testing fail. 

`ipassigner.NewIPAssigner(net.ParseIP("127.0.0.1"), dummyDeviceName)`

```
func NewIPAssigner(nodeIPAddr net.IP, dummyDeviceName string) (*ipAssigner, error) {
	_, egressInterface, err := util.GetIPNetDeviceFromIP(nodeIPAddr)
```

wenqiq commented:
I have tested the basic IPv6 SNAT function and put some info in my repo. https://github.com/wenqiq/antrea/blob/egress-ndp-demonstration/docs/egress-ipv6/egress-ipv6-practice.md.

Since last review, I found the NDP NA message received by external Node but not worked well, after debugging and verifying, I added the HopLimit parameter in sockInet6 which fixed the issue. `syscall.SetsockoptInt(sockInet6, syscall.IPPROTO_IPV6, syscall.IPV6_MULTICAST_HOPS, hopLimit)`

Also fixed the integration test case issue, the integration test seems back to normal.

PTAL. @tnqn

wenqiq commented:
/test-all

tnqn commented on test/integration/agent/ip_assigner_linux_test.go:
```diff
@@ -29,7 +31,29 @@ import (
+		}
+		return nil
+	}
+	nodeIP := getNodeIP()
```
There is a global variable nodeIP (defined in route_test.go) can be used directly.

wenqiq replied:
Done

tnqn commented on pkg/util/ip/ip.go:
```diff
@@ -179,3 +179,18 @@ func MustParseCIDR(cidr string) *net.IPNet {
+func MustIPv6(s string) net.IP {
+	ip := net.ParseIP(s)
+	if err := CheckIPv6(ip); err != nil {
+		return nil
```
AFAIK, a `MustDoSomething` function just "doSomething" and panics on error, which is used when the input must be valid, for example when it's used in testing or code with static input. 
https://github.com/kubernetes/kubernetes/blob/f840bd719dcfcc42ea2eb5fcdb025ea0dc0b9dfe/staging/src/k8s.io/apimachinery/pkg/util/runtime/runtime.go#L169
https://github.com/kubernetes/kubernetes/blob/f840bd719dcfcc42ea2eb5fcdb025ea0dc0b9dfe/pkg/kubelet/cm/cpuset/cpuset.go#L279

Can you make the function same style?
```
import utilnet "k8s.io/utils/net"

func MustParseIPv6(s string) net.IP {
	ip := net.ParseIP(s)
	if !utilnet.IsIPv6(ip) {
		panic(fmt.Errorf("invalid IPv6 address: %s", s))
	}
	return ip
}
```

wenqiq replied:
Done

tnqn reviewed (commented):
The commit message needs some adjustment.
1. The title should be clearer about what it does and can be shorter to fit most git tools: `Send NDP NA message upon assigning egress IP`
2. The body should be wrapped to ~76 characters (per line), which will look neater from git log. And it shouldn't just say there is a new package ndp.

wenqiq commented:
/test-all

wenqiq commented on pkg/agent/util/ndp/ndp_linux_test.go:
```diff
@@ -0,0 +1,77 @@
+				return
+			}(tt.ipv6Addr, tt.hardwareAddr); err != nil || !reflect.DeepEqual(got, tt.want) {
+				t.Errorf("NDPmarshalMessage() = %v, want %v", got, tt.want)
+			}
```
Done

tnqn reviewed (commented):
LGTM, commit message has few grammer issue.

> Implement Neighbor Discovery Protocol(NDP) neighbor advertisement for Egress IPv6 support,
once an IPv6 IP address has assigned to Node, an unsolicited Neighbor Advertisement ICMPv6
multicast packet will be sent, announcing the IP to all IPv6 nodes as per RFC4861.

Implement Neighbor Discovery Protocol(NDP) neighbor advertisement for Egress IPv6 support.
Once an IPv6 IP address has been assigned to Node, an unsolicited Neighbor Advertisement ICMPv6
multicast packet will be sent, announcing the IP to all IPv6 nodes as per RFC4861.

wenqiq commented:
Rebased upstream main branch and fixed grammer issues in commit message.
/test-all

tnqn reviewed (dismissed):
LGTM, will wait for one day before merging in case anyone wants to take another look

tnqn commented:
/test-all

tnqn commented:
/test-ipv6-all

tnqn commented:
@wenqiq I just notice the tests still don't run on IPv6 clusters. Can you have a following-up PR to enable egress tests on IPv6 and dual-stack clusters? otherwise we wouldn't know when it's broken by accident.

tnqn commented:
"Go / Check tidy, code generation and manifest (pull_request)" failed, which may be because go 1.17 update. Please update your local go version and regenerate the code.

wenqiq commented:
> @wenqiq I just notice the tests still don't run on IPv6 clusters. Can you have a following-up PR to enable egress tests on IPv6 and dual-stack clusters? otherwise we wouldn't know when it's broken by accident.

Sure, will implement it.

> "Go / Check tidy, code generation and manifest (pull_request)" failed, which may be because go 1.17 update. Please update your local go version and regenerate the code.

It seems there is no generated code or manifest changes in this PR.

wenqiq commented:
Rebased upstream/main.
/test-all

tnqn commented:
/test-ipv6-all

tnqn commented:
/test-ipv6-only-all

tnqn commented:
/test-e2e

wenqiq commented:
It seems `jenkins-ipv6-only-e2e`  test failed? How can we get the test result details?

tnqn commented:
I checked the IPv6 e2e tests failed on go build issue, which may be related to Go version upgrade. Since we haven't enabled egress test on IPv6, it shouldn't be affected in theory. I'm going to merge this. @wenqiq will have a follow-up PR for enabling egress test on IPv6 cluster.
